src/js/server/kettleValidation.js

```diff
@@ -0,0 +1,107 @@
+    var gssSchema = fluid.get(kettleValidator, ["options", "requestSchemas", schemaKey]);
+    var transformationRules = fluid.get(kettleValidator, ["options", "requestContentToValidate", rulesKey]);
+    var toValidate = fluid.model.transformWithRules(requestHandler.req, transformationRules);
+    var validationResults = globalValidator.validate(gssSchema, toValidate, schemaHash);
+
+    if (validationResults.isValid) {
+        requestHandler.handleValidRequest(requestHandler);
+    }
+    else {
+        var localisedErrors = gpii.schema.validator.localiseErrors(validationResults.errors, toValidate, kettleValidator.model.messages, kettleValidator.options.localisationTransform);
+        var localisedPayload = fluid.copy(validationResults);
+        localisedPayload.errors = localisedErrors;
+
+        var failurePayload = fluid.extend({}, kettleValidator.options.errorTemplate, localisedPayload);
+        requestHandler.events.onError.fire(failurePayload);
+    }
+};
+
+gpii.schema.kettle.validator.cacheSchemas = function (kettleValidator, globalValidator) {
+    fluid.each(kettleValidator.options.requestSchemas, function (requestSchema, schemaKey) {
+        var schemaHash = gpii.schema.hashSchema(requestSchema);
+        kettleValidator.schemaHashes[schemaKey] = schemaHash;
+        globalValidator.cacheSchema(requestSchema, schemaHash);
+    });
+};
+
+fluid.defaults("gpii.schema.kettle.validator", {
+    gradeNames: ["fluid.modelComponent"],
+    members: {
+        schemaHashes: {}
+    },
+    model: {
+        messages: gpii.schema.messages.validationErrors
+    },
+    localisationTransform: {
+        "": ""
+    },
+    errorTemplate: {
+        // "Bad Request": https://developer.mozilla.org/nl/docs/Web/HTTP/Status/400
+        statusCode: 400,
+        message: "Your request was invalid.  See the errors for details."
+    },
+    requestSchemas: {
+        default: {
+            "$schema": "gss-v7-full#"
+        }
+    },
+    requestContentToValidate: {
+        default: "{that}.options.requestContentToValidate.body",
+        body: {
+            "": "body"
+        },
+        params: {
+            "": "params"
+        },
+        query: {
+            "": "query"
+        }
+    },
+    invokers: {
+        validateRequest: {
+            funcName: "gpii.schema.kettle.validator.validateRequest",
+            args: ["{that}", "{gpii.schema.validator}", "{arguments}.0"] // kettleValidator, globalValidator, request
+        }
+    },
+    listeners: {
+        "onCreate.cacheSchemas": {
+            funcName: "gpii.schema.kettle.validator.cacheSchemas",
+            args:     ["{that}", "{gpii.schema.validator}"] // globalValidator
+        }
+    }
+});
+
+fluid.defaults("gpii.schema.kettle.app", {
+    gradeNames: ["kettle.app"],
+    components: {
+        validator: {
+            type: "gpii.schema.kettle.validator"
+        }
+    }
+});
+
+fluid.defaults("gpii.schema.kettle.request.http", {
```

Comment:
I think, as agreed, for ultimate performance we need to entirely axe the schema-specialised request grade. Requests will opt in to validation by making a reference to statically available middleware hosted in the "eternal".

Reply:
So, I have been reading up further and looking at the docs and tests in kettle.  If I understand you correctly, we would stuff a bunch of usable-as-middleware instances, one per type of data at the app level, and then refer to them as part of the `requestMiddleware stack` within a kettle.request.http grade.  Similar to [what's done here](https://github.com/fluid-project/kettle/blob/670396acbf4be31be009b2b2dee48373134ea94d/tests/shared/SessionTestDefs.js#L64) but without hosting the middleware component in the kettle.request grade itself.

Reply:
Yes, that's correct